syntax = "proto3";

package ai_service_rate_limiting;

service RateLimitService {
    // User management
    rpc CreateUser (CreateUserRequest) returns (CreateUserResponse);
    rpc GetUser (GetUserRequest) returns (GetUserResponse);
    rpc DeleteUser (DeleteUserRequest) returns (DeleteUserResponse);
    rpc ListUsers (ListUsersRequest) returns (ListUsersResponse);

    // Policy group management
    rpc CreatePolicyGroup (CreatePolicyGroupRequest) returns (CreatePolicyGroupResponse);
    rpc GetPolicyGroup (GetPolicyGroupRequest) returns (GetPolicyGroupResponse);
    rpc UpdatePolicyGroup (UpdatePolicyGroupRequest) returns (UpdatePolicyGroupResponse);
    rpc DeletePolicyGroup (DeletePolicyGroupRequest) returns (DeletePolicyGroupResponse);
    rpc ListPolicyGroups (ListPolicyGroupsRequest) returns (ListPolicyGroupsResponse);
    
    // Policy management
    rpc CreatePolicy (CreatePolicyRequest) returns (CreatePolicyResponse);
    rpc GetPolicy (GetPolicyRequest) returns (GetPolicyResponse);
    rpc UpdatePolicy (UpdatePolicyRequest) returns (UpdatePolicyResponse);
    rpc DeletePolicy (DeletePolicyRequest) returns (DeletePolicyResponse);
    rpc ListPolicies (ListPoliciesRequest) returns (ListPoliciesResponse);
    
    // Policy assignments
    rpc AssignPoliciesToUser (AssignPoliciesRequest) returns (AssignPoliciesResponse);
    rpc RemovePoliciesFromUser (RemovePoliciesRequest) returns (RemovePoliciesResponse);
    
    // Usage tracking
    rpc GetUserUsage (UserUsageRequest) returns (UserUsageResponse);
}

// --- Policy Group Management ---
message CreatePolicyGroupRequest {
    string name = 1;
}

message CreatePolicyGroupResponse {
    bool success = 1;
}

message GetPolicyGroupRequest {
    string policy_group_id = 1;
}

message GetPolicyGroupResponse {
    string policy_group_id = 1;
    string name = 2;
}

message UpdatePolicyGroupRequest {
    string policy_group_id = 1;
    string name = 2;
}

message UpdatePolicyGroupResponse {
    bool success = 1;
}

message DeletePolicyGroupRequest {
    string policy_group_id = 1;
}

message DeletePolicyGroupResponse {
    bool success = 1;
}

message ListPolicyGroupsRequest {
    int32 page = 1;
    int32 page_size = 2;
}

message ListPolicyGroupsResponse {
    repeated GetPolicyGroupResponse policy_groups = 1;
    int32 total_count = 2;
}

// --- User Management ---
message CreateUserRequest {
    string user_id = 1;
    string name = 2;
    string policy_group_id = 3;
}

message CreateUserResponse {
    bool success = 1;
}

message GetUserRequest {
    string user_id = 1;
}

message GetUserResponse {
    string user_id = 1;
    string name = 2;
    string policy_group_id = 3;
    repeated string policies = 4;
}

message DeleteUserRequest {
    string user_id = 1;
}

message DeleteUserResponse {
    bool success = 1;
}

message ListUsersRequest {
    int32 page = 1;
    int32 page_size = 2;
}

message ListUsersResponse {
    repeated GetUserResponse users = 1;
    int32 total_count = 2;
}

// --- Policy Management ---
message CreatePolicyRequest {
    string name = 1;
    string model = 2;
    string provider = 3;
    int32 input_token_limit = 4;
    int32 output_token_limit = 5;
    string period = 6;
    string policy_group_id = 7;
}

message CreatePolicyResponse {
    bool success = 1;
}

message GetPolicyRequest {
    string policy_id = 1;
}

message GetPolicyResponse {
    string name = 1;
    string model = 2;
    string provider = 3;
    int32 input_token_limit = 4;
    int32 output_token_limit = 5;
    string period = 6;
    string policy_group_id = 7;
}

message UpdatePolicyRequest {
    string policy_id = 1;
    string name = 2;
    string model = 3;
    string provider = 4;
    int32 input_token_limit = 5;
    int32 output_token_limit = 6;
    string period = 7;
    string policy_group_id = 8;
}

message UpdatePolicyResponse {
    bool success = 1;
}

message DeletePolicyRequest {
    string policy_id = 1;
}

message DeletePolicyResponse {
    bool success = 1;
}

message ListPoliciesRequest {
    int32 page = 1;
    int32 page_size = 2;
}

message ListPoliciesResponse {
    repeated GetPolicyResponse policies = 1;
    int32 total_count = 2;
}

// --- Policy Assignments ---
message AssignPoliciesRequest {
    string user_id = 1;
    repeated string policy_ids = 2;
}

message AssignPoliciesResponse {
    bool success = 1;
}

message RemovePoliciesRequest {
    string user_id = 1;
    repeated string policy_ids = 2;
}

message RemovePoliciesResponse {
    bool success = 1;
}

// --- Usage Tracking ---
message UserUsageRequest {
    string user_id = 1;
}

message UserUsageResponse {
    repeated UserUsageEntry usage = 1;
}

message UserUsageEntry {
    string model = 1;
    string provider = 2;
    int32 input_tokens_used = 3;
    int32 output_tokens_used = 4;
    int32 input_token_limit = 5;
    int32 output_token_limit = 6;
}
